### DISCUSSION TITLE
Bucket4j and infinispan

### DISCUSSION THREAD

**Discussion by git-volver** [vladimir-bukhtoyarov/bucket4j#366] [Q&A] [answered]
Hi,
Congratulations about this superb API.
I am in trouble to use it in my project, and would be glad for some help.
The infinispan API is already been used in my web project in "Embedded mode" with wildfly12 and java8. So, there is a infinispan_config.xml with a "cache-container" element used globally in application.
Can I put another  context-initializer element with class="io.github.bucket4j.grid.infinispan.serialization.Bucket4jProtobufContextInitializer" 
together the other existing in there?
And the obvious difficulty for me is how to use the code in the example you suggested in
[https://bucket4j.com/8.2.0/toc.html#bucket4j-infinispan](https://bucket4j.com/8.2.0/toc.html#bucket4j-infinispan)

`3.4.4. Example of Bucket instantiation
org.infinispan.functional.FunctionalMap.ReadWriteMap<K, byte[]> map = ...;
private static final InfinispanProxyManager<K> proxyManager = new InfinispanProxyManager(map);

...
BucketConfiguration configuration = BucketConfiguration.builder()
    .addLimit(Bandwidth.simple(1_000, Duration.ofMinutes(1)))
    .build(key, configuration);

Bucket bucket = proxyManager.builder().build(configuration);`

how to instantiate the map variable?

My intention is to control the rate limit from IP to certain request paths (like public domains) inside a servlet filter.

Cheers, and thanks by advanced.

**Comment by vladimir-bukhtoyarov** [comment, marked as answer]
Hello @git-volver 

> Can I put another context-initializer element with class="io.github.bucket4j.grid.infinispan.serialization.Bucket4jProtobufContextInitializer"
together the other existing in there?

According to Infinispan documentation it is possible https://infinispan.org/docs/stable/titles/encoding/encoding.html#registering-sci-embedded-caches_marshalling
![image](https://user-images.githubusercontent.com/6761830/231180966-83a882a8-4904-43d9-bde8-81b5be3d511b.png)

---- 
> how to instantiate the map variable?

If you have an instance of ```org.infinispan.Cache``` then initialization of finctional map is trivial:
```java
private static ReadWriteMap<String, byte[]> toMap(Cache<String, byte[]> cache) {
        FunctionalMapImpl<String, byte[]> functionalMap = FunctionalMapImpl.create(cache.getAdvancedCache());
        return ReadWriteMapImpl.create(functionalMap);
}
```
See this test if something unclear according to map instantiation https://github.com/bucket4j/bucket4j/blob/master/bucket4j-infinispan-all/bucket4j-infinispan/src/test/java/io/github/bucket4j/grid/infinispan/InfinispanTest.java
